File: objectiveai/objectiveai/src/chat/completions/request/provider.rs
//! Provider preferences for chat completion requests.

use serde::{Deserialize, Serialize};

/// Provider routing and selection preferences.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct Provider {
    /// Whether to allow providers to collect data.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data_collection: Option<ProviderDataCollection>,
    /// Whether to use zero data retention providers only.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub zdr: Option<bool>,
    /// How to sort/prioritize providers.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sort: Option<ProviderSort>,
    /// Maximum price constraints.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_price: Option<ProviderMaxPrice>,
    /// Preferred minimum throughput (tokens/second).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preferred_min_throughput: Option<f64>,
    /// Preferred maximum latency (seconds).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preferred_max_latency: Option<f64>,
    /// Hard minimum throughput requirement (tokens/second).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_throughput: Option<f64>,
    /// Hard maximum latency requirement (seconds).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_latency: Option<f64>,
}

/// Data collection policy for providers.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProviderDataCollection {
    /// Do not allow data collection.
    Deny,
    /// Allow data collection.
    Allow,
}

/// How to sort/prioritize providers.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProviderSort {
    /// Prioritize by price (cheapest first).
    Price,
    /// Prioritize by throughput (fastest first).
    Throughput,
    /// Prioritize by latency (lowest first).
    Latency,
}

/// Maximum price constraints per token type.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct ProviderMaxPrice {
    /// Maximum price per prompt token.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt: Option<rust_decimal::Decimal>,
    /// Maximum price per completion token.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completion: Option<rust_decimal::Decimal>,
    /// Maximum price per image.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image: Option<rust_decimal::Decimal>,
    /// Maximum price per audio second.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio: Option<rust_decimal::Decimal>,
    /// Maximum price per request.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request: Option<rust_decimal::Decimal>,
}

File: objectiveai-js/src/chat/completions/request/provider.ts
import z from "zod";
import { convert, type JSONSchema } from "../../../json_schema";

export const ProviderDataCollectionSchema = z
  .enum(["allow", "deny"])
  .describe("Specifies whether to allow providers which collect data.");
export type ProviderDataCollection = z.infer<
  typeof ProviderDataCollectionSchema
>;
export const ProviderDataCollectionJsonSchema: JSONSchema = convert(
  ProviderDataCollectionSchema,
);

export const ProviderSortSchema = z
  .enum(["price", "throughput", "latency"])
  .describe("Specifies the sorting strategy for provider selection.");
export type ProviderSort = z.infer<typeof ProviderSortSchema>;
export const ProviderSortJsonSchema: JSONSchema = convert(ProviderSortSchema);

export const ProviderMaxPriceSchema = z.object({
  prompt: z
    .number()
    .optional()
    .nullable()
    .describe("Maximum price for prompt tokens."),
  completion: z
    .number()
    .optional()
    .nullable()
    .describe("Maximum price for completion tokens."),
  image: z
    .number()
    .optional()
    .nullable()
    .describe("Maximum price for image generation."),
  audio: z
    .number()
    .optional()
    .nullable()
    .describe("Maximum price for audio generation."),
  request: z
    .number()
    .optional()
    .nullable()
    .describe("Maximum price per request."),
});
export type ProviderMaxPrice = z.infer<typeof ProviderMaxPriceSchema>;
export const ProviderMaxPriceJsonSchema: JSONSchema = convert(
  ProviderMaxPriceSchema,
);

export const ProviderSchema = z
  .object({
    data_collection: ProviderDataCollectionSchema.optional().nullable(),
    zdr: z
      .boolean()
      .optional()
      .nullable()
      .describe(
        "Whether to enforce Zero Data Retention (ZDR) policies when selecting providers.",
      ),
    sort: ProviderSortSchema.optional().nullable(),
    max_price: ProviderMaxPriceSchema.optional().nullable(),
    preferred_min_throughput: z
      .number()
      .optional()
      .nullable()
      .describe("Preferred minimum throughput for the provider."),
    preferred_max_latency: z
      .number()
      .optional()
      .nullable()
      .describe("Preferred maximum latency for the provider."),
    min_throughput: z
      .number()
      .optional()
      .nullable()
      .describe("Minimum throughput for the provider."),
    max_latency: z
      .number()
      .optional()
      .nullable()
      .describe("Maximum latency for the provider."),
  })
  .describe("Options for selecting the upstream provider of this completion.")
  .meta({ title: "CompletionProvider" });
export type Provider = z.infer<typeof ProviderSchema>;
export const ProviderJsonSchema: JSONSchema = convert(ProviderSchema);
